internal/api/option/error.go

```diff
@@ -0,0 +1,91 @@
+package option
+
+import (
+	"context"
+	"net/http"
+
+	"connectrpc.com/connect"
+	"github.com/pkg/errors"
+	kubeerr "k8s.io/apimachinery/pkg/api/errors"
+)
+
+var (
+	_ connect.Interceptor = &errorInterceptor{}
+)
+
+type errorInterceptor struct{}
+
+func newErrorInterceptor() connect.Interceptor {
+	return &errorInterceptor{}
+}
+
+func (i *errorInterceptor) WrapUnary(next connect.UnaryFunc) connect.UnaryFunc {
+	return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
+		res, err := next(ctx, req)
+		if err != nil {
+			return nil, i.wrapError(err)
+		}
+		return res, nil
+	}
+}
+
+func (i *errorInterceptor) WrapStreamingClient(
+	next connect.StreamingClientFunc,
+) connect.StreamingClientFunc {
+	// TODO: Support streaming client when necessary
+	return next
+}
+
+func (i *errorInterceptor) WrapStreamingHandler(
+	next connect.StreamingHandlerFunc,
+) connect.StreamingHandlerFunc {
+	return func(ctx context.Context, conn connect.StreamingHandlerConn) error {
+		if err := next(ctx, conn); err != nil {
+			return i.wrapError(err)
+		}
+		return nil
+	}
+}
+
+func (i *errorInterceptor) wrapError(err error) error {
+	var connectErr *connect.Error
+	if ok := errors.As(err, &connectErr); ok {
+		return err
+	}
+	var statusErr *kubeerr.StatusError
+	if ok := errors.As(err, &statusErr); ok {
+		return connect.NewError(httpStatusToConnectCode(statusErr.Status().Code), statusErr)
+	}
+	return err
```

Comment:
If we get to here, don't we want to default to an "internal server error" code?

Reply:
Connect will care about this case - ([docs](https://connectrpc.com/docs/go/errors/#working-with-errors))

> Connect handlers attach status codes to errors using the NewError function. Handlers should return coded errors; if they don't, Connect will use the deadline_exceeded code for context.DeadlineExceeded, canceled for context.Canceled, and unknown for all other errors.

And other interceptor _may_ affect status code as we did like `errorInterceptor` - so I think we can just return raw error

Reply:
> and unknown for all other errors

This is the sticky point that keeps coming up.

It is illuminating, to see the documentation you linked to and finally realize (💡) that "unknown" may be _conventional_ or at least _common_ for connect when dealing with unanticipated errors.

When you zoom out and stop looking at the implementation detail that is connect, I still find it doesn't make good sense. Being unanticipated isn't the same as being unknown. We'll always know what happened. It will be in the logs.

> And other interceptor may affect status code

Our only other interceptor is the log interceptor and it does not affect status.

If we add another that _may_ affect status, we would have to be careful about that, but isn't it pretty unlikely that we would do so? It's not unlikely that we'd add another interceptor. I think it's unlikely that we'd add one that obscures the error received from the handler.